Forgot password

kevwu27 opened the pull request:
## Tracking Info

Resolves #20, #21 

## Changes

### **Sign In:**
- added error states (incorrect email/password message, not valid email message)
- when signing in, if user has NOT verified their email, will display an error state that allows user to resend the verification email
- else if user HAS verified their email and logs in, if it is their first time logging in, it will take them to the "/" page (placeholder for the intro video page for now) and if not, it will take them to the "/signup" page (placeholder for the home page)
- Added a firstLogin field to user objects in MongoDB, keeps track of whether it's the user's first time logging in.
- More API routes created: now can query/get users by email, and can update user objects in MongoDB -- as of now, only purpose is to change the firstLogin field
- Created an /auth route. This page will deal with email verification and password resetting. This is because for these operations, firebase only allows one unique link for them, so we created an /auth page as an intermediary page that will simply display 1) any errors risen throughout the process, 2) act as a "loading" screen in transition (for a split second though), and 3) direct users to the /emailverified or /resetpassword page.
- Reorganized code so that if the backend is done or MongoDB does not successfully create a user but firebase was successful, the firebase account that was created is removed, so it's as if the account was never made. The verification email is also not sent until after it is confirmed that MongoDB object creation is successful. -- changes in CreateAccountPanel.tsx

### **Forgot Password:**
- all links between pages are now linked properly
- on the forgot password page, when user clicks "Send Link", firebase sends a link to the email for resetting their password
- Goes through /auth page, similar to email verification
- created /resetpassword page, including the error states, making sure that the password is at least 6 characters and that both passwords entered match before being able to hit "Reset"
- Password will be updated through firebase Auth, user will now be able to sign with this new password 

## Testing

- tested through localhost
- tried to replicate the error states
- went through making an account, resetting password, verifying email, etc. (pictures below)
- User should NOT be able to sign in unless they verified their email
- If first time signing in, taken to the "/" page
- Else taken to the "/signup" page
- Resetting password sends a link to email to /resetpassword page, upon resetting, signing in now works with the new password

## Confirmation of Change
Can test out by starting up both frontend and backend servers, and observing firebase and mongoDB objects. 

### **Sign In:**
Invalid account or email/password incorrect
![image](https://github.com/user-attachments/assets/4adc2c0f-e2ed-460a-9b80-08b0c043b590)

Email is invalid
![image](https://github.com/user-attachments/assets/4852a9e7-8663-498e-9cbb-66d28c341624)

Signing in, but email is not yet verified
<img width="356" alt="Screenshot 2025-02-20 at 12 57 35 AM" src="https://github.com/user-attachments/assets/dcff61e5-16ab-4fe4-9080-d5303a6b905f" />

Email verification:
https://github.com/user-attachments/assets/7c1a4724-003a-4876-a15f-3c25bf98b637

### **Forgot Password:**

Resetting password
https://github.com/user-attachments/assets/e8916722-b873-45b8-910c-48e3228b6bb6

OrigamiStarz reviewed (changes requested):
Looking great so far! I haven't fully looked over the code or tested everything you mentioned, but something I noticed is that when the user tries the "forgot password" page, inputs an email not in the database, and clicks the resend link, the console outputs "Success!" despite no email being sent. This seems to only occur when the resend link is clicked, and not when the user sees the "Email Sent!" page, and it happens regardless of whether the BE server is down or not.

<img width="1440" alt="Screenshot 2025-02-21 at 12 29 26 AM" src="https://github.com/user-attachments/assets/2b398b9f-59e5-433d-9654-b1513ebffd10" />

eshaan-s18 reviewed (changes requested):
Great job on this guys! Appreciate all your hard work as always. Here's a couple bugs I noticed to fix:

1. for some reason the sign up page is not sending verification emails to the right email. In these screenshots, i first signed up for an account with essharma@ucsd.edu, and it correctly sent me a verification email. then after i signed up for an account with eshaansharma18@gmail.com, and it incorrectly sent me a verification email to essharma@ucsd.edu

<img width="711" alt="Screenshot 2025-02-21 at 6 47 48 PM" src="https://github.com/user-attachments/assets/8976fff7-1840-4f17-a683-200a1ce9e8bf" />
<img width="560" alt="Screenshot 2025-02-21 at 6 48 15 PM" src="https://github.com/user-attachments/assets/c2698cbc-c900-4eb0-942c-c3e7601d2bad" />
<img width="420" alt="Screenshot 2025-02-21 at 6 48 22 PM" src="https://github.com/user-attachments/assets/9b5133e0-0c45-4ce2-8cef-2003029a18fd" />



Based on this i think something is wrong with how the email for the verification is being set and updated–and i believe this will address the issue we were seeing from the email being blank before. In past commits this functionality was working properly, so see if anything got changed! I believe it is sending the verification email to the right email based on my testing, however lets double check that. The displayed email might be the only thing that is wrong if thats the case.


2. For the `Forgot Password` page, lets add an error box for if they try to send a reset password email to an email that does not have an account-just as an extra safety check.
<img width="548" alt="Screenshot 2025-02-21 at 6 53 09 PM" src="https://github.com/user-attachments/assets/0da33578-b583-4e61-9108-0b4a4ed7a12c" />

i think that's all for now that I noticed. Great job on this otherwise! Once you make these changes ill do some additional testing. Thanks for your hard work guys! Let me know when you make these changes and if you have any questions!

kevwu27 commented:
Previously the email displayed on the verification page was based on the email stored on localStorage, I changed it so that it simply passes the data between panels. 

Also added an error state to ForgotPassword if no account exists within the MongoDB.

<img width="1436" alt="Screenshot 2025-02-24 at 4 18 18 PM" src="https://github.com/user-attachments/assets/97bafe3a-7850-4daf-9cc3-a2669909fa65" />

eshaan-s18 reviewed (approved):
Sorry for the delay on this guys, this looks good to me!

OrigamiStarz reviewed (changes requested):
Good job on everything! I see that everything we mentioned earlier were fixed. 

Just two things that happened while I was testing: 
- There is no indication of having clicked the "Click to Resend" in the Email Sent! page. I accidentally clicked it too many times, and because of latency (?), I ended up getting ~20 emails in the span of a minute after some delay. I think an indicator would be nice (and also, good job on blocking more emails after that, it does not seem too spammable despite this). 
- It is possible to change to change your account password to the same password (could try implementing password history, i.e. not letting the user use the same password if they have use it in their last 5 passwords, but I feel like this can be its own PR so it's not really necessary).

TLDR: Could you guys add an indication for when the "Click to Send" is clicked in the Email Sent page? Also, sorry for reviewing it so late.

shashwatb14 reviewed (approved):
Everything seems good to me! So sorry for the late review

Sweekrit-B commented:
Added more functionality to the error states. Tells you when the email is resent and how long to wait before retrying
![image](https://github.com/user-attachments/assets/98c33be5-7c44-4cec-af07-ab04c8b22f34)
and makes sure that the "Too many requests error" does not stay.

In regards to the password storage, made a Slack GC because that might be its own PR.

eshaan-s18 reviewed (approved):
Ok this looks good to me now. Great work guys! Approved and merging
